src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs

```diff
@@ -325,6 +325,28 @@ internal virtual NamedTypeSymbol LookupMetadataType(ref MetadataTypeName emitted
                 }
             }
 
+            if (isTopLevel && MetadataHelpers.DecodeFileType(emittedTypeName.UnmangledTypeName) is (not -1 and var ordinal, var typeName))
+            {
+                // also do a lookup for file types from source.
+                namespaceOrTypeMembers = scope.GetTypeMembers(typeName);
+                foreach (var named in namespaceOrTypeMembers)
+                {
+                    if (named is SourceMemberContainerTypeSymbol { AssociatedSyntaxTree: SyntaxTree tree }
+                        && named.DeclaringCompilation.GetSyntaxTreeOrdinal(tree) == ordinal
```

Comment:
>named.DeclaringCompilation.GetSyntaxTreeOrdinal(tree) == ordinal [](http://example.com/codeflow?start=27&length=64)

If this method is ever called for a merged namespace we might run into some conflicts. We probably never do that, but let's add an assert at the top of the method to confirm.